lib/tests/filesystem.sh

```diff
@@ -38,11 +38,15 @@ mkfifo fifo
 checkPathType() {
     local path=$1
     local expectedPathType=$2
-    local actualPathType=$(nix-instantiate --eval --strict --json 2>&1 \
+    local result
+    stderr=$(mktemp)
+    if ! result=$(nix-instantiate --eval --strict --json 2>"$stderr" \
         -E '{ path }: let lib = import <nixpkgs/lib>; in lib.filesystem.pathType path' \
-        --argstr path "$path")
-    if [[ "$actualPathType" != "$expectedPathType" ]]; then
-        die "lib.filesystem.pathType \"$path\" == $actualPathType, but $expectedPathType was expected"
+        --argstr path "$path"); then
+        result=$(<"$stderr")
```

Comment:
Not happy about conflating the out and err output. I think having two implementations should be cleaner.

Also `!` is not great. In Nix we have [`expectStderr`](https://github.com/NixOS/nix/blob/940e9eb8dd6deacb2e41aca91d65bd0e616011d6/tests/common/vars-and-functions.sh.in#L204-L216) for the use case of checking error messages. It checks the exit code explicitly and redirects stderr for you.

Reply:
I think this code here is fairly neat, because it concisely implements checking stdout when the exit code was 0, but stderr when it was non-0, without conflating the two. And this works in all the cases in this file, there's never a need to check stderr for exit code 0 or the other way around.

I can't quite see how `expectStderr` would work here.

Reply:
The caller would have to be explicit about expecting a failure, which is good. This could be done by having two functions (eg `expectValue` vs `expectFailure`) or an argument for the expected exit code. `expectFailure` should probably have that as a parameter anyway; might as well use `0` to represent the `expectValue` case.
To be fair conflating errors and values is unlikely to lead to actual problems, but code that's explicit about its single intent is easier to understand. I also think most people would expect failures to be tested differently compared to "return values", so this clever solution is a bit surprising.